src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp

```diff
@@ -72,6 +72,45 @@ void CgroupV1Controller::set_subsystem_path(char *cgroup_path) {
             buf[MAXPATHLEN-1] = '\0';
             _path = os::strdup(buf);
           }
+        } else {
+          // Find the longest common substring as namespaces are hierarchical.
+          // If there are no matches between the root and the cgroup path,
+          // the controller's mount point will be used, which is a reasonable
+          // fallback.
+          int buflen;
+          strncpy(buf, _mount_point, MAXPATHLEN);
+          buf[MAXPATHLEN-1] = '\0';
+          buflen = strlen(buf);
+          if ((buflen + strlen(cgroup_path)) > (MAXPATHLEN-1)) {
+              return;
+          }
+          char tmp_root_buf[MAXPATHLEN+1];
+          char tmp_cgroup_buf[MAXPATHLEN+1];
+          strcpy(tmp_root_buf, _root);
+          strcpy(tmp_cgroup_buf, cgroup_path);
+          char* tmp_root = &tmp_root_buf[0];
+          char* tmp_cgroup = &tmp_cgroup_buf[0];
+          char* root_save = NULL;
+          char* cgroup_save = NULL;
+          char* root_token = NULL;
+          char* cg_token = NULL;
+          char* buf_ptr = buf + buflen;
+          while ((root_token = strtok_r(tmp_root, "/", &root_save)) != NULL &&
+                 (cg_token = strtok_r(tmp_cgroup, "/", &cgroup_save)) != NULL) {
+            if (root_token == NULL || cg_token == NULL) {
+              break;
+            }
+            if (strcmp(root_token, cg_token) == 0) {
+              strncpy(buf_ptr, "/", 2);
+              buf_ptr++;
+              strcpy(buf_ptr, root_token);
+              buf_ptr = buf_ptr + strlen(root_token);
+            }
+            tmp_root = NULL;
+            tmp_cgroup = NULL;
+          }
+          buf[MAXPATHLEN-1] = '\0';
+          _path = os::strdup(buf);
```

Comment:
I think this code can be simplified a lot with stringStream and without strtok, so no need for fixed buffers (which may fail with longer path names) and no need for writable string copies on the stack.

Something like this:
```
stringStream ss;
ss.print_raw(_mount_point);
const char* p1 = _root;
const char* p2 = cgroup_path;
int last_matching_dash_pos = -1;
for (int i = 0; *p1 == *p2 && *p1 != 0; i ++) {
	if (*p1 == '/') {
		last_matching_dash_pos = i;
	}
}
ss.print_raw(_root, last_matching_dash_pos);
// Now use ss.base() to access the assembled string
```

Reply:
Nice, thanks! I'll update it.

Reply:
Done.